tests/boards/native_posix/cpu_wait/src/main.c

```diff
@@ -0,0 +1,295 @@
+	/* @ 121ms */
+	k_sleep(Z_TIMEOUT_TICKS(2));
+	/* @ 150ms */
+	k_busy_wait(WASTED_TIME);
+}
+
+/**
+ * @brief Test the POSIX arch k_busy_wait and cpu_hold while another thread
+ * takes time during this test thread waits
+ *
+ * Note: This test relies on the exact timing of the ticks.
+ * For native_posix it works, with a tick of 10ms. In general this test will
+ * probably give problems if the tick time is not a relatively even number
+ * of microseconds
+ */
+static void test_cpu_hold_with_another_thread(void)
+{
+	uint64_t time2, time1;
+
+	k_sleep(Z_TIMEOUT_TICKS(1)); /* Wait until tick boundary */
+
+	/* Let TIME_WASTER thread start */
+	k_sem_give(&start_sema);
+
+	/* k_busy_wait part: */
+
+	time1 = posix_get_hw_cycle();
+	/* @40 ms */
+	k_busy_wait(2*ONE_TICK_TIME + 1);
+	/* The thread should have switched in and have used
+	 * WASTED_TIME us right after 2*one_tick_time
+	 * As that is longer than 1 tick + 1us, the total should be
+	 * 2 ticks + WASTE_TIME
+	 */
+
+	time2 = posix_get_hw_cycle();
+
+	zassert_true(time2 - time1 == 2*ONE_TICK_TIME + WASTED_TIME,
+			"k_busy_wait failed %llu-%llu != %u\n",
+			time2, time1, 2*ONE_TICK_TIME);
+
+	/* @61 ms */
+	k_sleep(Z_TIMEOUT_TICKS(1)); /* Wait almost 2 ticks */
+
+	time1 = posix_get_hw_cycle();
+	/* @80 ms */
+	k_busy_wait(ONE_AND_HALF_TICKS);
+	/* The thread should have used WASTED_TIME us after 1 tick,
+	 * but as that is lower than 1.5 ticks, in total the wait should be
+	 * 1.5 ticks
+	 */
+
+	time2 = posix_get_hw_cycle();
+
+	zassert_true(time2 - time1 == ONE_AND_HALF_TICKS,
+			"k_busy_wait failed %llu-%llu != %u\n",
+			time2, time1, ONE_TICK_TIME);
+
+	/* CPU hold part: */
+
+	/* @95 ms */
+	k_sleep(Z_TIMEOUT_TICKS(1)); /* Wait almost 2 ticks */
+
+	time1 = posix_get_hw_cycle();
+	/* @110 ms */
+	posix_cpu_hold(ONE_TICK_TIME + 1);
+	/* The thread should have used WASTED_TIME us after 1 tick,
+	 * so the total should be 1 tick + WASTED_TIME + 1 us.
+	 * That is compared to k_busy_wait() we spend 1 us more in this context
+	 */
+
+	time2 = posix_get_hw_cycle();
+
+	zassert_true(time2 - time1 == ONE_TICK_TIME + 1 + WASTED_TIME,
+			"k_busy_wait failed %llu-%llu != %u\n",
+			time2, time1, ONE_TICK_TIME);
+
+
+	/* @121.001 ms */
+	k_sleep(Z_TIMEOUT_TICKS(1)); /* Wait almost 2 ticks */
+
+	time1 = posix_get_hw_cycle();
+	/* @140 ms */
+	posix_cpu_hold(ONE_AND_HALF_TICKS);
+	/* The thread should have used WASTED_TIME us after 1 tick,
+	 * so the total wait should be 1.5 ticks + WASTED_TIME.
+	 * That is 1.5 ticks in this context, and WASTED_TIME in the other
+	 * thread context
+	 */
+
+	time2 = posix_get_hw_cycle();
+
+	zassert_true(time2 - time1 == ONE_AND_HALF_TICKS + WASTED_TIME,
+			"k_busy_wait failed %llu-%llu != %u\n",
+			time2, time1, ONE_TICK_TIME);
+}
+
+/**
+ * Replacement system tick timer interrupt handler which wastes time
+ * before calling the real one
+ */
+void np_timer_isr_test_replacement(const void *arg)
+{
+	ARG_UNUSED(arg);
+
+	k_busy_wait(WASTED_TIME);
```

Comment:
I believe something would break if `WASTED_TIME` >= tick time. Should there be a check somewhere to make sure that is the case? Could `WASTED_TIME` be expressed as a fraction of a tick?

Reply:
I guess we could add it. But the whole test timings are coupled very deeply. You cannot change any of them without breaking all of it.

Reply:
> You cannot change any of them without breaking all of it.

Which is all the more reason to be paranoid and check (statically) that the assumptions about tick time and/or `WASTED_TIME` hold.

Reply:
done